### DISCUSSION TITLE
How to wait to the completion of a write?

### DISCUSSION THREAD

**Discussion by zapta** [hbldh/bleak#1274] [Q&A]
I have a program that needs to perform a no-response write to a device and exit. What is a good way to wait for the completion of the write operation before existing?

```
        await self.__client.write_gatt_char(self.__my_chrc, bytearray([0x01]))
```

Currently I am doing this but it feels hackish.
```
    for i in range(1000):
        main_event_loop.run_until_complete(do_nothing())
        time.sleep(0.001)
    sys.exit(f"\nDevice updated")
```

**Comment by dlech** [comment]
If the characteristic supports it, do a write with response instead of write without response. If the device doesn't support it, then there is no way to wait.

**Reply by zapta** [reply]
How does the response=True works, does the write_gatt_char() return only when the response received from the device?  Also, how does the caller of write_gatt_char() accesses the response status?

**Reply by dlech** [reply]
> How does the response=True works, does the write_gatt_char() return only when the response received from the device?

That is a question best answered by reading a book about BLE. I didn't read it but this looks useful: https://iotexpert.com/ble-write-request-write-command-signed-write-command-prepare-write/



> Also, how does the caller of write_gatt_char() accesses the response status?

If there was an error, an exception will be raised. Getting the actual error code from the exception is currently platform-specific.

**Comment by zapta** [comment]
Thanks @dlech. I changed the call to be with response and now the call blocks until the response and and throws an error in case of a GATT error status, which is what I need.

In this case I don't care about the response's value so I am good. Would be nice though for the write function to return the gatt status and response value in case response=True. It will simplify the user's code and will abstract platform differences.
